### ISSUE TITLE
Any chance to support the .sh3d format from the Sweet Home 3D Software? -> wwww.sweethome3d.com

### ISSUE THREAD

**Issue by justinlimbo** [WARP-LAB/files_3dmodelviewer#3]

**Comment by kroko** [comment]
TL;DR

1) A better way to support `sh3d` IMHO is a separate application, that would mirror [Sweet Home 3D Online Manager](https://www.sweethome3d.com/SweetHome3DJSOnline.jsp) 
2) In order to view `sh3d` 3D model, three.js loader for `sh3d` is needed (either `sh3d` to _raw_ three.js structures or middleware that converts `sh3d` to already supported _format_).

---

My two cents.

- It seems that [Sweet Home 3D Online Manager](https://www.sweethome3d.com/SweetHome3DJSOnline.jsp) exists, which should cover "share/view/collaborate" online part if set up...
- ... which by itself is not an argument against supporting this format within this app/within Nextcloud ecosystem, but introduces questions -> should supposedly generic "open, standartised and broadly used file format viewer" which this claims to be 😆 support it /or should be separate Nextcloud app, and if it is supported what would be the technical pathway.

1) **The first suggestion** would be for you to approach devs (or maybe you are involved yourself) and ask, if they would be interested integrating it within Nextcloud. [Nextcloud Draw.io / Diagrams.net integration app](https://github.com/jgraph/drawio-nextcloud/) is a good example. The structure of such integration app is done by iframing existing Sweet Home 3D Online Manager and passing file URL (which would demand some coding) that has to be viewed/edited. If some backend is needed, it is theirs (and unless YOLO, then authorisation, rate limiting would be needed). There are some integrations apps such as [CADViewer](https://github.com/CADViewer/NextCloud/) which even ship [PHP + binary backend](https://github.com/CADViewer/NextCloud/blob/main/README.md). I mean, it is a non issue, if Nextcloud admin needs the support, the app can be "more than one click to install away". Stuff in Nextcloud like  Collabora, TURN also are not one click away.

2) **The second suggestion**. I took a quick look at <https://sourceforge.net/p/sweethome3d/code/HEAD/tree/trunk/SweetHome3DJS/>. HTML5 frontend + Java backend. I did not dig that deep in order to check if `sh3d` file is parsed and fed to GL buffers only using frontend (JavaScript) or backend is involved (Java), but from what I saw IMHO frontend is custom rolled WebGL stuff.
You could approach _Sweet Home 3D_ devs and ask them to create [three.js loader](https://threejs.org/docs/#manual/en/introduction/Loading-3D-models), that takes `sh3d` stream, parses it (they now the schema inside out) using JavaScript and represents it as three.js demands. That would make their file format accessible broadly including this app, as it uses three.js under the hood. Moreover, `sh3d` probably has exporters, that can be reworked to convert `sh3d` using JavaScript to [any of already supported formats in three.js](https://github.com/mrdoob/three.js/tree/dev/examples/jsm/loaders) (i.e.,`sh3d` to `obj+mtl`), producing _middleware_, as the resulting buffer could be then fed to the existing importer of choice. Note, that the context here is `sh3d` **viewer**, which allows _metadata_ loss.
